lib/pbench/server/api/resources/__init__.py

```diff
@@ -1577,19 +1590,79 @@ def _dispatch(
                     message=HTTPStatus.INTERNAL_SERVER_ERROR.phrase,
                 )
 
+        audit = None
+
+        if schema.operation is not OperationCode.READ:
+            user = Auth.token_auth.current_user()
+            dataset_found = schema.get_param_by_type(ParamType.DATASET, params)
+            dataset = dataset_found.value if dataset_found else None
+
+            audit = Audit.create(
+                operation=schema.operation,
+                status=AuditStatus.BEGIN,
+                user=user,
+                name=schema.audit_name,
+                dataset=dataset,
+                object_type=schema.audit_type,
+            )
+
+        # Pass the root audit object to the API implementation. Normally, we'll
+        # automatically generate an audit finalization; however if the API
+        # wants to emit a special audit sequence it can disable "finalize"
+        # in the context. It can also pass "attributes" by setting that
+        # field.
+        auditing = {
+            "audit": audit,
+            "finalize": bool(audit),
+            "status": AuditStatus.SUCCESS,
+            "reason": None,
+            "attributes": None,
+        }
+
         try:
-            return execute(params, request)
+            response = execute(params, request, {"auditing": auditing})
+            if auditing["finalize"]:
```

Comment:
We wouldn't create an audit entry if `auditing["finalize"]` attribute is `False`? Is it possible to have this attribute `False`?

Reply:
It's a capability I don't think I'm currently using; but, yeah, the idea is that if an API implementation wants to do something custom, it can suppress the default `SUCCESS` entry.

Reply:
Okay lets cross the bridge when we get there :+1: